jwst/cube_build/ifu_cube.py

```diff
@@ -1204,17 +1262,374 @@ def map_detector_to_outputframe(self, this_par1, this_par2,
+        wavelength range covered by the IFU. This is not the case for NIRSpec the
+        30 different slices map to different FOV on the range of wavelengths.
+
+        Paramteter
+        ---------
+        this_par1: Channel (MIRI) or Grating (NIRSpec)
+        coord1: xi coordinates of input data (~x coordinate in IFU space)
+        coord2: eta coordinates of input data (~y coordinate in IFU space)
+        wave: wavelength of input data
+        roiw_ave: average spectral roi used to determine which wavelength bins
+            the input values would be mapped to
+        slice: integer slice value of input data (used in MIRI case to find
+            the points of the edge slices.)
+        """
+
+        # MIRI mapping:
+        # The FOV is roughtly the same for all the wavelenth ranges.
+        # The offset in the slices makes the calculation of the four corners
+        # of the FOV more comlicated. So we only use the two slices at
+        # the edges of the FOV to define the 4 corners.
+        # Note we can not use the wcs.footprint (because this footprint only
+        # consists of 4 values  ra min, ra max, dec min, dec max and we
+        # need 4 corners made of 8 different values.
+
+        if self.instrument == 'MIRI':
+
+            # find the wavelength boundaries of the band - use two extreme slices
+            wavemin = np.amin(wave)
+            wavemax = np.amax(wave)
+            iwavemin = np.absolute(np.array(wavemin - self.zcoord) / (self.cdelt3_normal/2.0))
+            iwavemax = np.absolute(np.array(wavemax - self.zcoord) / (self.cdelt3_normal/2.0))
+            imin = np.where(iwavemin == np.amin(iwavemin))[0]
+            imax = np.where(iwavemax == np.amin(iwavemax))[0]
+
+            for w in range(imin[0], imax[0]):
+                wave_distance = np.absolute(self.zcoord[w] - wave)
+
+                # pull out the two extreme slices in the FOV.
+                # use these points to set the FOV
+                start_region = self.instrument_info.GetStartSlice(this_par1)
+                end_region = self.instrument_info.GetEndSlice(this_par1)
+
+                istart = start_region
+                while istart < end_region:
+
+                    index_use = np.where((wave_distance < roiw_ave) & (slice == istart))
+                    if len(index_use[0]) > 0:
+                        coord2_start = coord2[index_use]
+                        coord1_start = coord1[index_use]
+                        break
+                    istart = istart + 1
+
+                iend = end_region
+
+                while iend > start_region:
+                    index_use = np.where((wave_distance < roiw_ave) & (slice == iend))
+                    if len(index_use[0]) > 0:
+                        coord2_end = coord2[index_use]
+                        coord1_end = coord1[index_use]
+                        break
+                    iend = iend - 1
```

Comment:
Python loops (because of how slow they are) always make me wonder if the same operation can be done using clever array-based operations instead. But of course I don't understand enough of the details of exactly what's being done here to make any suggestions. It would probably take some experimentation.

Reply:
I do think this can be improved. I need the MIRI and NIRSpec team to do some experimentation and let me know about using the same dqflagging for all wavelengths in a band. Currently it is setting the dqflagging per wavelength bin. This might be overkill.